ftNode/ftnodeDriver.cpp

```diff
@@ -0,0 +1,432 @@
+{
+    size_t numberOfChannels = 0;
+    std::vector<double> measurements;
+};
+
+struct SerialPortWrenchData
+{
+    std::vector<double> forceBuffer;
+    bool forceUpdateFlag;
+
+    std::vector<double> torqueBuffer;
+    bool torqueUpdateFlag;
+
+};
+
+class ftnodeDriver::Impl
+{
+public:
+    mutable std::mutex mutex;
+    yarp::dev::ISerialDevice *iSerialDevice = nullptr;
+
+    std::string serialComPortName;
+
+    // Buffer for storing data from serial port
+    std::vector<SerialPortWrenchData> serialPortWrenchDataVector;
+
+    int numberOfFTSensors;
+    std::vector<std::vector<double>> wrenchScalingFactors;
+    AnalogSensorData analogSensorData;
+};
+
+
+// Default constructor
+ftnodeDriver::ftnodeDriver()
+    : PeriodicThread(period)
+    , pImpl{new Impl()}
+{}
+
+// Destructor
+ftnodeDriver::~ftnodeDriver() = default;
+
+bool ftnodeDriver::open(yarp::os::Searchable& config)
+{
+    // ==================================
+    // Check the configuration parameters
+    // ==================================
+
+    // Period of the this device
+    if (!(config.check("period") && config.find("period").isFloat64())) {
+        yInfo() << LogPrefix << "Using default period: " << period << "s";
+    }
+    else {
+        period = config.find("period").asFloat64();
+        yInfo() << LogPrefix << "Using the period : " << period << "s";
+    }
+
+    // The number of sensors to be configured for IAnalogSensor interface
+    // Currently, the serial port streams data from only one FT and in the
+    // future it will also stream the data from the four FTs of the FTshoes
+    if (!(config.check("numberOfFTSensors") && config.find("numberOfFTSensors").isInt())) {
+        yError() << LogPrefix << "Option 'channels' not found or not a valid integer";
+        return false;
+    }
+
+    // Check for wrench scaling factor group
+    yarp::os::Bottle& scalingFactorGroup = config.findGroup("WRENCH_SCALING_FACTOR");
+
+    if (scalingFactorGroup.isNull()) {
+        yError() << LogPrefix << "Failed to find SCALING_FACTOR group";
+        return false;
+    }
+
+    if (!(scalingFactorGroup.check("LeftFront") && scalingFactorGroup.find("LeftFront").isList())) {
+        yError() << LogPrefix << "Option LeftFront scaling factor list is not found";
+        return false;
+    }
+
+    if (!(scalingFactorGroup.check("LeftRear") && scalingFactorGroup.find("LeftRear").isList())) {
+        yError() << LogPrefix << "Option LeftRear scaling factor list is not found";
+        return false;
+    }
+
+    if (!(scalingFactorGroup.check("RightFront") && scalingFactorGroup.find("RightFront").isList())) {
+        yError() << LogPrefix << "Option RightFront scaling factor list is not found";
+        return false;
+    }
+
+    if (!(scalingFactorGroup.check("RightRear") && scalingFactorGroup.find("RightRear").isList())) {
+        yError() << LogPrefix << "Option RightRear scaling factor list is not found";
+        return false;
+    }
```

Comment:
those checks are specific for the shoes use-case so should be avoided. I think those checks can be safely removed since in the rest of the device we are never using those specific group names.